File: bit_map.c
/*
 * Implements of Bit Map
 * 
 * Author: Minchiuan Gao <mqgao@outlook.com>
 *
 *  The simpest implements of bit map is to write in char(because char only is one byte), but still waste 7/8 of the memory spacce.
 *  This programming will implement it by ** array of int **
 *  
 *  If we have a bit array A, A[0] will contain 4 bytes which are 32 bits.
 *  
 *  So if we need a bitmap which length of 10, 000, 000, which apporiximatly is 1M, 2^20, so we need a length is 2^20 / 2^5 = 2^15 = 2 ^ 5 * 2 ^ 10 = 
 *  10 * 1k = 10k ~= 10, 000. It means we need an A[30000]
 *  
 *  Elements with index * k * in the bit map is :
 *		+ bit position k % 32
 *		+ array elements with index A[k/32]
 */

void setBit(int A[], int k){
	int pos = k % 32;
	int index = k / 32;
	unsigned int flag = 1; // flag = 0000000000000000000000001
	flag = flag << pos; // flag =	 0000000001000000000000000
	A[i] = A[i] | flag; // set the positino of A[i] to 1
	/* the shorten version
	A[k/32] |= 1 << (k%32); //Set the bit at the k-th position in A[i]
	*/
}

void clearBit(int A[], int k){
	int index = k / 32;
	int pos = k % 32;
	unsigned int flag = 1; // flag = 0000000000000000000001
	flag = flag << pos;   //  flag = 0000000000000100000000
	flag =~ flag;         //  flag = 1111111111111011111111
	A[i] = A[i] & flag; // set the position in A[i] to zero
	/*
	 * the short version
	 *		A[k/32] &= ~(1<< (k%32));
	 */
}

int testBit(int A[], int k){
	int i = k / 32;
	int pos = k % 32;
	unsigned int flag = 1;

	flag = flag << pos;
	if (A[i] & flag){ // 1 and 1 is 1
		return 1;	
	}else{ // 0 and 1 is 0, other and 0 also is 0;
		return 0; 
	}
}

File: bit_map.py
"""
* bit map implemented by python
* Author: Minchiuan Gao (mqgao@outlook.com)
"""


class BitVector:
    def __init__(self, length):
        self.bit_array = [0] * (length/32)

    def set_bit(self, number):
        index = number / 32
        position = number % 32

        flag = 1
        flag = flag << position
        self.bit_array[index] |= flag

    def clear_bit(self, number):
        index = number / 32
        position = number % 32
        flag = 1

        flag <<= position
        flag = ~flag  # 1111111111111111110111111111111111111
        self.bit_array[index] &= flag

    def test_bit(self, number):
        index = number / 32
        position = number % 32
        flag = 1

        flag = flag << position

        if self.bit_array[index] & flag != 0:
            return 1
        else:
            return 0


def test():
    total_number = 100000000
    bitVector = BitVector(total_number)

    number_file = file('number.data', 'r')

    for line in number_file.xreadlines():
        number = int(line.rstrip())
        # print line.rstrip()
        if bitVector.test_bit(number) == 0:
            bitVector.set_bit(number)
        else:
            print(' dulipcated number ' + line)

    sorted_file = file('sorted.data', 'w+')
    for i in range(total_number):
            sorted_file.write(str(i) + '\n')


def python_sort():
    array = []
    number_file = file('random.data', 'r')
    for line in number_file.xreadlines():
        array.append(int(line.rstrip()))

    array = sorted(array)
    sorted_file = file('system_sorted_file.data', 'w+')
    for i in array:
        sorted_file.write(str(i)+'\n')


if __name__ == '__main__':
    import time

    start = time.time()
    test()
    print 'used time == ' + str(time.time() - start)

